authorizations/authorization-ranger/src/main/java/org/apache/gravitino/authorization/ranger/RangerAuthorizationPlugin.java

```diff
@@ -0,0 +1,1015 @@
+                                                            .contains(role.name())) {
+                                                          policyItem.getRoles().add(role.name());
+                                                        }
+                                                      }
+                                                    });
+                                            try {
+                                              rangerClient.updatePolicy(policy.getId(), policy);
+                                            } catch (RangerServiceException e) {
+                                              throw new RuntimeException(e);
+                                            }
+                                          });
+                                });
+                      });
+            });
+
+    return execResult.get();
+  }
+
+  /**
+   * Because one Ranger policy maybe contain multiple Gravitino securable objects, <br>
+   * So we need to find the corresponding policy item mapping to remove the user.
+   */
+  @Override
+  public Boolean onRevokedRolesFromUser(List<Role> roles, User user) throws RuntimeException {
+    roles.stream()
+        .forEach(
+            role -> {
+              checkRangerRole(role.name());
+              GrantRevokeRoleRequest grantRevokeRoleRequest =
+                  createGrantRevokeRoleRequest(role.name(), user.name(), null);
+              try {
+                rangerClient.revokeRole(rangerServiceName, grantRevokeRoleRequest);
+              } catch (RangerServiceException e) {
+                // Ignore exception to support idempotent operation
+              }
+            });
+
+    return Boolean.TRUE;
+  }
+
+  /**
+   * Grant the roles to the group. <br>
+   * 1. Create a group in the Ranger if the group is not exist. <br>
+   * 2. Create a role in the Ranger if the role is not exist. <br>
+   * 3. Add this group to the role. <br>
+   * 4. Add the role to the policy item base the metadata object. <br>
+   *
+   * @param roles The roles to grant to the group.
+   * @param group The group to grant the roles.
+   */
+  @Override
+  public Boolean onGrantedRolesToGroup(List<Role> roles, Group group) throws RuntimeException {
+    // If the group is not exist, then create it.
+    onGroupAdded(group);
+
+    AtomicReference<Boolean> execResult = new AtomicReference<>(Boolean.TRUE);
+    roles.stream()
+        .forEach(
+            role -> {
+              createRangerRoleIfNotExists(role.name());
+              GrantRevokeRoleRequest grantRevokeRoleRequest =
+                  createGrantRevokeRoleRequest(role.name(), null, group.name());
+              try {
+                rangerClient.grantRole(rangerServiceName, grantRevokeRoleRequest);
+              } catch (RangerServiceException e) {
+                // Ignore exception to support idempotent operation
+              }
+
+              role.securableObjects().stream()
```

Comment:
like `grantRolesToUser `

Reply:
Yes, you are right.
We need support idempotent operation. May be user doesn't call `grantRolesToUser` function.

Reply:
We should remove this policy modification, keep align with `grantRoleToUser`